agent/src/com/thoughtworks/go/agent/AgentController.java

```diff
@@ -17,231 +17,106 @@
-            LOG.info("About to get cookie from the server.");
-            String cookie = server.getCookie(agentIdentifier(), agentRuntimeInfo.getLocation());
-            agentRuntimeInfo.setCookie(cookie);
-            LOG.info("Got cookie: {}", cookie);
-        }
     }
 
-    private void handleIfSecurityException(Exception e) {
-        if (!isCausedBySecurity(e)) {
-            return;
-        }
+    private void handleSecurityException(Exception e) {
         sslInfrastructureService.invalidateAgentCertificate();
         LOG.error("There has been a problem with one of Go's SSL certificates." +
-                        " This can be caused by a man-in-the-middle attack, or by pointing the agent to a new server, or by" +
+                        " This can be caused by a man-in-the-middle attack, or by pointing the agent to a new e, or by" +
                         " deleting and re-installing Go Server. Go will ask for a new certificate. If this" +
                         " fails to solve the problem, try deleting config/trust.jks in Go Agent's home directory.",
                 e);
     }
 
-    void retrieveWork() {
-        AgentIdentifier agentIdentifier = agentIdentifier();
-        LOG.debug("[Agent Loop] {} is checking for work from Go", agentIdentifier);
-        Work work;
-        try {
-            agentRuntimeInfo.idle();
-            work = server.getWork(agentRuntimeInfo);
-            if (!(work instanceof NoWork)) {
-                LOG.debug("[Agent Loop] Got work from server: [{}]", work.description());
-            }
-            runner = new JobRunner();
-            runner.run(work, agentIdentifier, server, manipulator, agentRuntimeInfo, packageAsRepositoryExtension, scmExtension, taskExtension);
-        } catch (UnregisteredAgentException e) {
-            LOG.warn("[Agent Loop] Invalid agent certificate with fingerprint {}. Registering with server on next iteration.", e.getUuid());
-            sslInfrastructureService.invalidateAgentCertificate();
-        } finally {
-            agentRuntimeInfo.idle();
-        }
+    protected abstract void work() throws Exception;
+
+    protected AgentRegistry getAgentRegistry() {
+        return agentRegistry;
     }
 
-    public void executeAgentInstruction() {
-        if (systemEnvironment.isWebsocketEnabled()) {
-            return;
-        }
-        if (runner != null) {
-            runner.handleInstruction(instruction, agentRuntimeInfo);
-        }
+    protected AgentAutoRegistrationProperties getAgentAutoRegistrationProperties() {
+        return agentAutoRegistrationProperties;
+    }
+
+    protected SystemEnvironment getSystemEnvironment() {
+        return systemEnvironment;
+    }
+
+    protected AgentIdentifier agentIdentifier() {
+        return new AgentIdentifier(SystemUtil.getLocalhostNameOrRandomNameIfNotFound(), SystemUtil.getClientIp(systemEnvironment.getServiceUrl()), agentRegistry.uuid());
```

Comment:
Hostname and IP computations have been inlined here. Given that this method is called during every ping and loop call, does it make sense to determine this once and reuse thereafter?